### PR TITLE
refactor: add client request type

### PR DESCRIPTION
**Pull request by agostbiro**
Add client request type as the first step towards https://github.com/NomicFoundation/edr/issues/256

I ended up naming the client request type `RequestMethod` instead of the discussed `Request`, as the latter conflicts with the name of the struct that represents a JSON-RPC request (version + method + id) in the client.

### CODE REVIEW

File: crates/edr_eth/src/remote/client.rs [hunk 1]
```diff
@@ -89,10 +86,10 @@ pub enum RpcClientError {
     },
 
     /// A response is missing from a batch request.
-    #[error("Missing response for method: '{method_invocation:?}' for request id: '{id:?}' in batch request")]
+    #[error("Missing response for method: '{method:?}' for request id: '{id:?}' in batch request")]
     MissingResponse {
         /// The method invocation for which the response is missing.
-        method_invocation: MethodInvocation,
+        method: Box<RequestMethod>,
```
Comment (agostbiro) [inline crates/edr_eth/src/remote/client.rs:92]: I had to box this as clippy was complaining about a `Result<SerializedRequest, RpcClientError>` being 152 bytes large which is too big. The weird thing is that I checked, and before this PR, its size was 408 bytes.

File: crates/edr_eth/src/remote/cacheable_method_invocation.rs [hunk 2]
```diff
@@ -86,193 +72,127 @@ enum CacheableMethodInvocation<'a> {
...
-            MethodInvocation::NetVersion(_) => Ok(CacheableMethodInvocation::NetVersion),
+            RequestMethod::NetVersion(_) => Ok(CacheableRequestMethod::NetVersion),
 
             // Explicit to make sure if a new method is added, it is not forgotten here.
-            MethodInvocation::Accounts(_)
-            | MethodInvocation::BlockNumber(_)
-            | MethodInvocation::Call(_, _, _)
-            | MethodInvocation::Coinbase(_)
-            | MethodInvocation::EstimateGas(_, _)
-            | MethodInvocation::FeeHistory(_, _, _)
```
Comment (Wodann) [inline crates/edr_eth/src/remote/cacheable_method_invocation.rs:246]: We need to keep the `FeeHistory` for the implementation of `eth_FeeHistory`. See `node.ts` in Hardhat

File: crates/edr_eth/src/remote/request_methods.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+use crate::{
+    remote::{eth::GetLogsInput, BlockSpec, PreEip1898BlockSpec},
+    U256,
+};
+
+/// Methods for requests to a remote Ethereum node. Only contains methods
+/// supported by the [`crate::remote::client::RpcClient`].
+#[derive(Clone, Debug, PartialEq, serde::Serialize)]
+#[serde(tag = "method", content = "params")]
+pub enum RequestMethod {
```
Comment (Wodann) [inline crates/edr_eth/src/remote/request_methods.rs:12]: Missing `feeHistory`

**Review by Wodann — COMMENTED** [review]
Overall LGTM. There is just one missing client request method that we'll need in the implementation of `eth_feeHistory`

File: crates/edr_eth/src/remote/cacheable_method_invocation.rs (hunk 2 above)
Comment (agostbiro) [reply crates/edr_eth/src/remote/cacheable_method_invocation.rs:246]: Thanks, fixed in: [8c20657](https://github.com/NomicFoundation/hardhat/pull/4687/commits/8c20657c1125d4f3efba6b012f7b1d9c9d2889f3)

File: crates/edr_eth/src/remote/request_methods.rs (hunk 3 above)
Comment (agostbiro) [reply crates/edr_eth/src/remote/request_methods.rs:12]: Fixed in: [8c20657](https://github.com/NomicFoundation/hardhat/pull/4687/commits/8c20657c1125d4f3efba6b012f7b1d9c9d2889f3)

**Review by alcuadrado — APPROVED** [review]

**Pull request merged**